pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -356,44 +360,84 @@ func validatePipelineParameterVariables(tasks []PipelineTask, params []ParamSpec
 		}
 	}
 
-	return validatePipelineVariables(tasks, "params", parameterNames, arrayParameterNames)
+	return validatePipelineParametersVariables(tasks, "params", parameterNames, arrayParameterNames)
 }
 
-func validatePipelineVariables(tasks []PipelineTask, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
+func validatePipelineParametersVariables(tasks []PipelineTask, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
 	for _, task := range tasks {
-		for _, param := range task.Params {
-			if param.Value.Type == ParamTypeString {
-				if err := validatePipelineVariable(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, paramNames); err != nil {
-					return err
-				}
-				if err := validatePipelineNoArrayReferenced(fmt.Sprintf("param[%s]", param.Name), param.Value.StringVal, prefix, arrayParamNames); err != nil {
+		if err := validateParametersVariablesInParameters(task.Params, prefix, paramNames, arrayParamNames); err != nil {
+			return err
+		}
+		if err := validateParametersVariablesInWhenExpressions(task.WhenExpressions, prefix, paramNames, arrayParamNames); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validateParametersVariablesInWhenExpressions(whenExpressions WhenExpressions, prefix string, paramNames sets.String, arrayParamNames sets.String) *apis.FieldError {
+	name, path := "task when expression", "pipelinespec.when"
+	for _, we := range whenExpressions {
+		if err := validatePipelineStringVariable(fmt.Sprintf("when.input[%s]", we.Input), we.Input, prefix, name, path, paramNames, arrayParamNames); err != nil {
```

Comment:
Thanks a bunch @jerop for fixing it in this PR 🙏 and sorry for circling back, but you have done an amazing job separating and refactoring param validation and when expression validation, I would like to see one more change here. This `when` is not needed here, it adds extra `when` in the overall message: `pipelinespec.when.when.input[...])` and on line 385 causes `pipelinespec.when.when.values[...])`

Reply:
good catch, fixed it :)